### PR TITLE
Return pipe status in response header

### PR DESCRIPTION
**Pull request by sebastianlavender**

### CODE REVIEW

File: pipe-storage-sqlite/src/main/java/com/tesco/aqueduct/pipe/storage/sqlite/SQLiteStorage.java [hunk 1]
```diff
@@ -160,14 +178,26 @@ public void write(final Message message) {
...
                 statement.setLong(2, offset.getValue().getAsLong());
                 statement.setLong(3, offset.getValue().getAsLong());
                 statement.execute();
             });
     }
 
+    @Override
+    public void write(PipeState pipeState) {
+        execute(SQLiteQueries.UPSERT_PIPE_STATE,
+                ((connection, statement) -> {
```
Comment (LorenzoCianciaruso) [inline pipe-storage-sqlite/src/main/java/com/tesco/aqueduct/pipe/storage/sqlite/SQLiteStorage.java]: shift indentation back once

File: pipe-http-server-cloud/src/integration/groovy/PipeCloudServerIntegrationSpec.groovy [hunk 2]
```diff
@@ -31,7 +31,10 @@ class PipeCloudServerIntegrationSpec extends Specification {
         sql = new Sql(pg.embeddedPostgres.postgresDatabase.connection)
 
         dataSource = Mock()
-        dataSource.connection >> sql.connection
+        dataSource.connection >>> [
+            new Sql(pg.embeddedPostgres.postgresDatabase.connection).connection,
+            new Sql(pg.embeddedPostgres.postgresDatabase.connection).connection
```
Comment (adam-yorwerth) [inline pipe-http-server-cloud/src/integration/groovy/PipeCloudServerIntegrationSpec.groovy]: How come we need 2 connections here?

File: pipe-http-server/src/main/java/com/tesco/aqueduct/pipe/http/PipeReadController.java [hunk 3]
```diff
@@ -79,7 +79,10 @@ public PipeStateResponse state(@Nullable final List<String> type) {
         final long retryTime = messageResults.getRetryAfterSeconds();
 
         LOG.debug("pipe read controller", String.format("set retry time to %d", retryTime));
-        MutableHttpResponse<List<Message>> response = HttpResponse.ok(list).header(HttpHeaders.RETRY_AFTER, String.valueOf(retryTime));
+        MutableHttpResponse<List<Message>> response = HttpResponse.ok(list)
+            .header(HttpHeaders.RETRY_AFTER, String.valueOf(retryTime))
+            .header(HttpHeaders.PIPE_STATE,
+                pipeStateProvider.getState(types, messageReader).isUpToDate() ? UP_TO_DATE.toString() : OUT_OF_DATE.toString());
```
Comment (adam-yorwerth) [inline pipe-http-server/src/main/java/com/tesco/aqueduct/pipe/http/PipeReadController.java]: Formatting

File: pipe-storage-sqlite/src/main/java/com/tesco/aqueduct/pipe/storage/sqlite/SQLiteStorage.java [hunk 4]
```diff
@@ -47,6 +48,12 @@ private void createOffsetTableIfNotExists() {
             (connection, statement) -> statement.execute());
     }
 
+    private void createPipeStateTableIfNotExists() {
+        execute(
+            SQLiteQueries.PIPE_STATE_TABLE,
+            (Connection, statement) -> statement.execute());
```
Comment (adam-yorwerth) [inline pipe-storage-sqlite/src/main/java/com/tesco/aqueduct/pipe/storage/sqlite/SQLiteStorage.java]: ); can go onto a new line

File: pipe-storage-sqlite/src/main/java/com/tesco/aqueduct/pipe/storage/sqlite/SQLiteStorage.java [hunk 5]
```diff
@@ -72,7 +79,18 @@ public MessageResults read(final List<String> types, final long offset, final St
...
+
+    private PipeState getPipeState() {
+        return executeGet(SQLiteQueries.GET_PIPE_STATE,
+            (connection, statement) -> {
+                try (ResultSet resultSet = statement.executeQuery()) {
+                    return resultSet.next()
+                            ? PipeState.valueOf(resultSet.getString("value"))
+                            : PipeState.OUT_OF_DATE;
+                }
+            });
```
Comment (adam-yorwerth) [inline pipe-storage-sqlite/src/main/java/com/tesco/aqueduct/pipe/storage/sqlite/SQLiteStorage.java]: Formatting

File: pipe-storage-sqlite/src/main/java/com/tesco/aqueduct/pipe/storage/sqlite/SQLiteStorage.java [hunk 6]
```diff
@@ -160,14 +178,25 @@ public void write(final Message message) {
...
+    @Override
+    public void write(PipeState pipeState) {
+        execute(SQLiteQueries.UPSERT_PIPE_STATE,
+            ((connection, statement) -> {
+                statement.setString(1, "pipe_state");
+                statement.setString(2, pipeState.toString());
+                statement.setString(3, pipeState.toString());
+                statement.execute();
+            }));
+    }
```
Comment (adam-yorwerth) [inline pipe-storage-sqlite/src/main/java/com/tesco/aqueduct/pipe/storage/sqlite/SQLiteStorage.java]: formatting

**Pull request merged**